src/code/mempak.c

```diff
@@ -1,184 +1,266 @@
+        osSyncPrintf("mempak: write %d byte '%c' (%d)->%d\n", size, letter,
+                     sMempakFiles[MEMPAK_LETTER_TO_INDEX(letter)], error);
     }
-    PadMgr_ReleaseSerialEventQueue(&gPadMgr, mq);
-
+    PadMgr_ReleaseSerialEventQueue(&gPadMgr, serialEventQueue);
     return ret;
 }
 
-s32 Mempak_Read(s32 controllerNb, char idx, void* buffer, s32 offset, s32 size) {
-    OSMesgQueue* mq;
+/**
+ * Reads data from the file identified with `letter`.
+ *
+ * @param controllerNum Unused, the controller used is that which was last passed to `Mempak_Init`
+ * @param letter Memory pak file letter, in the range 'A' to ('A' + MEMPAK_MAX_FILES)
+ * @param buffer Buffer to read data into
+ * @param offset Offset into the file to read from
+ * @param size Size in bytes
+ * @return true if the operation completed successfully, false otherwise
+ */
+s32 Mempak_Read(s32 controllerNum, char letter, void* buffer, s32 offset, s32 size) {
+    OSMesgQueue* serialEventQueue;
     s32 error;
     s32 ret = false;
     s32 pad;
 
-    mq = PadMgr_AcquireSerialEventQueue(&gPadMgr);
+    serialEventQueue = PadMgr_AcquireSerialEventQueue(&gPadMgr);
 
     if (size < sMempakFreeBytes) {
-        error = osPfsReadWriteFile(&sMempakPfsHandle, sMempakFiles[idx - 'A'], 0, offset, size, buffer);
+        error = osPfsReadWriteFile(&sMempakPfsHandle, sMempakFiles[MEMPAK_LETTER_TO_INDEX(letter)], PFS_READ, offset,
+                                   size, buffer);
         if (error == 0) {
             ret = true;
         }
-        osSyncPrintf("mempak: read %d byte '%c' (%d)<-%d\n", size, idx, sMempakFiles[idx - 'A'], error);
+        osSyncPrintf("mempak: read %d byte '%c' (%d)<-%d\n", size, letter, sMempakFiles[MEMPAK_LETTER_TO_INDEX(letter)],
+                     error);
     }
-    PadMgr_ReleaseSerialEventQueue(&gPadMgr, mq);
+    PadMgr_ReleaseSerialEventQueue(&gPadMgr, serialEventQueue);
     return ret;
 }
 
-s32 Mempak_Alloc(s32 controllerNb, char* idx, s32 size) {
-    OSMesgQueue* mq;
+/**
+ * Creates a new file on the memory pak.
+ *
+ * @param controllerNum Unused, the controller used is that which was last passed to `Mempak_Init`
+ * @param letter Memory pak file letter, in the range 'A' to ('A' + MEMPAK_MAX_FILES).
+ *      If this points to a valid file letter the new file will be created using that letter, otherwise it will create
+ *      a file using the first free letter and return it through this argument
+ * @param size File size
+ * @return true if the operation completed successfully, false otherwise
+ */
+s32 Mempak_CreateFile(s32 controllerNum, char* letter, s32 size) {
+    OSMesgQueue* serialEventQueue;
     s32 error;
-    s32 ret = 0;
+    s32 ret = false;
     s32 i;
     s32 pad;
 
-    mq = PadMgr_AcquireSerialEventQueue(&gPadMgr);
+    serialEventQueue = PadMgr_AcquireSerialEventQueue(&gPadMgr);
+
+    if (*letter >= MEMPAK_INDEX_TO_LETTER(0) && *letter <= MEMPAK_INDEX_TO_LETTER(MEMPAK_MAX_FILES)) {
```

Comment:
... why is `*letter == MEMPAK_INDEX_TO_LETTER(MEMPAK_MAX_FILES)` ok?

Note `s32 sMempakFiles[11];` instead of 10 matches (I haven't investigated further)